internal/retry/retry.go

```diff
@@ -49,9 +52,21 @@ func (r *IntervalRetrier) Do(ctx context.Context) error {
 			return err
 		}
 
+		// Check for context expiration.
 		select {
 		case <-ctx.Done():
-			return ctx.Err()
+			if lastErr == nil {
+				// The context already expired after the first attempt, so we just return the error we have.
+				lastErr = err
+			}
+			return fmt.Errorf("retries failed before %w: last err: %w", ctx.Err(), lastErr)
+		default:
+			lastErr = err
```

Comment:
My idea was to set this e.g. in `case <-ticker.C():` below and change `return fmt.Errorf("retries failed before %w: last err: %w", ctx.Err(), lastErr)` to `return fmt.Errorf("retries failed before %w: err: %w last err: %w", ctx.Err(), err, lastErr)` This way we also should return all interesting error, right?

Reply:
I also thought of simply `errors.Join`ing everything, but at the end of the day this error is directly user visible so I'd at least try to reduce the amount of repetition. In the case of `ctx.Done`, what's the chance that the last error is anything other than `ctx.Err` (potentially wrapped)?

Reply:
If ctx is done then I expect the err to the a wrapped instance of the ctx error. The lastErr is the more interesting error, I think, since it shows why the requests until that point weren't successful.